File: frontend/assets/ts/core/openai/capabilityChecks.ts
/* SoAI - Shared OpenAI capability checks [frontend/assets/ts/core/openai/capabilityChecks.ts] */
// SPDX-License-Identifier: LicenseRef-SoAI-Source-1.0

import { isBoolean, isNumber, isString } from '@core/typeGuards.ts';
import type { ModelData } from '@core/types/modelTypes.ts';
import type { OpenAICategory } from '@core/openai/capabilityCategories.ts';
import { isJsonArray, isJsonObject, type JsonObject, type JsonValue } from '@core/types/jsonValues.ts';

interface OpenAIEndpointCapabilityEntry {
    openaiCapabilities?: JsonValue | null | undefined;
}

const normalizeOpenAICapabilityToken = (value: string): string => value.trim().toLowerCase();

const normalizeCapabilityBoolean = (value: JsonValue | undefined): boolean => {
    if (isBoolean(value)) {
        return value;
    }
    if (isNumber(value)) {
        return value !== 0;
    }
    if (isString(value)) {
        const lowered = value.trim().toLowerCase();
        if (!lowered) {
            return false;
        }
        if (lowered === '1' || lowered === 'true' || lowered === 'yes' || lowered === 'on') {
            return true;
        }
        if (lowered === '0' || lowered === 'false' || lowered === 'no' || lowered === 'off') {
            return false;
        }
        return true;
    }
    return Boolean(value);
};

const isOpenAICapabilityEnabled = (openai: JsonObject | null | undefined, category: OpenAICategory, token: string): boolean => {
    if (!isJsonObject(openai)) {
        return false;
    }
    const normalizedToken = normalizeOpenAICapabilityToken(token);
    if (!normalizedToken) {
        return false;
    }
    const flatValue = openai[normalizedToken];
    if (flatValue !== undefined) {
        return normalizeCapabilityBoolean(flatValue);
    }
    const section = openai[category];
    if (!isJsonObject(section)) {
        return false;
    }
    if (!(normalizedToken in section)) {
        return false;
    }
    return normalizeCapabilityBoolean(section[normalizedToken]);
};

const hasOpenAIModality = (modalities: JsonValue | undefined | null, token: string): boolean => {
    if (!isJsonArray(modalities)) {
        return false;
    }
    const normalizedToken = normalizeOpenAICapabilityToken(token);
    if (!normalizedToken) {
        return false;
    }
    for (const entry of modalities) {
        if (!isString(entry)) {
            continue;
        }
        if (normalizeOpenAICapabilityToken(entry) === normalizedToken) {
            return true;
        }
    }
    return false;
};

const supportsOpenAIInputFeatureForModel = (model: ModelData | null, modalityToken: string, capabilityToken: string): boolean => {
    if (!model) {
        return false;
    }
    if (!hasOpenAIModality(model.modalities ?? null, modalityToken)) {
        return false;
    }
    const capabilities = model.openaiCapabilities;
    if (!isJsonObject(capabilities)) {
        return true;
    }
    return isOpenAICapabilityEnabled(capabilities, 'chat_features', capabilityToken);
};

const supportsVisionInputForModel = (model: ModelData | null): boolean => supportsOpenAIInputFeatureForModel(model, 'vision', 'vision');

const supportsOpenAIEndpointForEntry = (entry: OpenAIEndpointCapabilityEntry | null, endpointToken: string, defaultWhenCapabilitiesMissing = false): boolean => {
    if (!entry) {
        return false;
    }
    const capabilities = entry.openaiCapabilities;
    if (!isJsonObject(capabilities)) {
        return defaultWhenCapabilitiesMissing;
    }
    return isOpenAICapabilityEnabled(capabilities, 'endpoints', endpointToken);
};

const supportsOpenAIEndpointForModel = (model: ModelData | null, endpointToken: string): boolean => supportsOpenAIEndpointForEntry(model, endpointToken);

export { hasOpenAIModality, isOpenAICapabilityEnabled, normalizeOpenAICapabilityToken, supportsOpenAIEndpointForEntry, supportsOpenAIEndpointForModel, supportsVisionInputForModel };
export type { OpenAICategory };

File: backend/core/openai/capability_checks.py
"""SoAI - OpenAI capability predicate helpers [backend/core/openai/capability_checks.py]"""
# SPDX-License-Identifier: LicenseRef-SoAI-Source-1.0

from __future__ import annotations

from collections.abc import Mapping
from typing import TYPE_CHECKING

from core.validation.boolean_coercion import coerce_bool_with_default

if TYPE_CHECKING:
    from core.types.json import JSONValue

__all__ = (
    "has_openai_modality",
    "has_openai_model_modality",
    "is_openai_audio_input_supported",
    "is_openai_capability_enabled",
    "is_openai_input_feature_supported",
    "is_openai_model_audio_input_supported",
    "is_openai_model_capability_enabled",
    "is_openai_model_vision_input_supported",
    "is_openai_vision_input_supported",
    "normalize_openai_capability_token",
)


def normalize_openai_capability_token(token: str) -> str:
    return str(token or "").strip().lower()


def is_openai_capability_enabled(
    openai_capabilities: Mapping[str, JSONValue],
    category: str,
    token: str,
) -> bool:
    normalized_token = normalize_openai_capability_token(token)
    if not normalized_token:
        return False
    flat_value = openai_capabilities.get(normalized_token)
    if flat_value is not None:
        return coerce_bool_with_default(flat_value, default=False, strict=False)
    nested_value = openai_capabilities.get(category)
    if isinstance(nested_value, Mapping):
        return coerce_bool_with_default(
            nested_value.get(normalized_token),
            default=False,
            strict=False,
        )
    return False


def has_openai_modality(modalities: JSONValue, token: str) -> bool:
    normalized_token = normalize_openai_capability_token(token)
    if not normalized_token or not isinstance(modalities, list):
        return False
    for entry in modalities:
        if not isinstance(entry, str):
            continue
        if normalize_openai_capability_token(entry) == normalized_token:
            return True
    return False


def is_openai_input_feature_supported(
    *,
    modalities: JSONValue,
    openai_capabilities: JSONValue,
    modality_token: str,
    capability_token: str,
) -> bool:
    if not has_openai_modality(modalities, modality_token):
        return False
    if not isinstance(openai_capabilities, Mapping):
        return True
    return is_openai_capability_enabled(
        openai_capabilities,
        "chat_features",
        capability_token,
    )


def is_openai_vision_input_supported(
    *,
    modalities: JSONValue,
    openai_capabilities: JSONValue,
) -> bool:
    return is_openai_input_feature_supported(
        modalities=modalities,
        openai_capabilities=openai_capabilities,
        modality_token="vision",
        capability_token="vision",
    )


def is_openai_audio_input_supported(
    *,
    modalities: JSONValue,
    openai_capabilities: JSONValue,
) -> bool:
    return is_openai_input_feature_supported(
        modalities=modalities,
        openai_capabilities=openai_capabilities,
        modality_token="audio",
        capability_token="input_audio",
    )


def is_openai_model_capability_enabled(
    model_entry: Mapping[str, JSONValue],
    *,
    category: str,
    token: str,
) -> bool:
    openai_capabilities = model_entry.get("openai_capabilities")
    if not isinstance(openai_capabilities, Mapping):
        return False
    return is_openai_capability_enabled(openai_capabilities, category, token)


def has_openai_model_modality(
    model_entry: Mapping[str, JSONValue],
    *,
    token: str,
) -> bool:
    return has_openai_modality(model_entry.get("modalities"), token)


def is_openai_model_vision_input_supported(model_entry: Mapping[str, JSONValue]) -> bool:
    return is_openai_vision_input_supported(
        modalities=model_entry.get("modalities"),
        openai_capabilities=model_entry.get("openai_capabilities"),
    )


def is_openai_model_audio_input_supported(model_entry: Mapping[str, JSONValue]) -> bool:
    return is_openai_audio_input_supported(
        modalities=model_entry.get("modalities"),
        openai_capabilities=model_entry.get("openai_capabilities"),
    )
